tests/run-tests-parallel.py

```diff
@@ -0,0 +1,120 @@
+#
+# 3. Neither the name of the copyright holder nor the names of its contributors may be
+#    used to endorse or promote products derived from this software without specific
+#    prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+# THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+# THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+# The parallelism celing can be controlled with MONERO_PARALLEL_TEST_JOBS env variable,
+# for example:
+#
+# MONERO_PARALLEL_TEST_JOBS=1     tests/run-tests-parallel.py
+# MONERO_PARALLEL_TEST_JOBS=nproc tests/run-tests-parallel.py
+#
+# The minimum between the ceiling and a currently selected reasonable number of threads is used in the end.
+# The reasonable number is selected as a number, that delivers the solution in the shortest time.
+"""
+
+import os
+import sys
+import subprocess
+import multiprocessing
+from timeit import default_timer as timer
+
+NUM_PROC_MAX_REASONABLE=2 # This might be increased, once the core_tests are divided into many more independent pieces or simply sped up
+TESTS_EXCLUDED_REGEX="unit_tests|functional_tests_rpc" # These tests collide with core_tests
+ERR_CODE=1
+
+def get_forced_ceiling():
+    key = 'MONERO_PARALLEL_TEST_JOBS'
+    if key not in os.environ:
+        job_ceiling = multiprocessing.cpu_count()
+        print("No parallelism ceiling selected. Defaulting to nproc, so", job_ceiling)
+    else:
+        val = os.environ[key]
+        if val == "nproc":
+            job_ceiling=multiprocessing.cpu_count()
+        else:
+            job_ceiling = int(val)
+            
+        print("Parallelism ceiling selected. Using", job_ceiling, "jobs.")
+        
+    return job_ceiling
+
+def run_fail_fast(num_proc_final):
+	# Run the excluded tests first, then everything but excluded.
+	# In the current situation, the excluded tests are the most probable and quickest to fail, giving an early feedback.
+    ctest1 = get_ctest_command(num_proc_final, '-R')
+    ctest2 = get_ctest_command(num_proc_final, '-E')
+    
+    ret1 = run_cmd(ctest1)
+    if ret1 != 0:
+        return ret # Fail fast here
+    ret2 = run_cmd(ctest2)
+    if ret2 != 0:
+        return ret
+    
+    return 0
+
+def run_all():
+    # Run all of the tests, no matter if the first part fails
+    ctest1 = get_ctest_command(num_proc_final, '-R')
```

Comment:
These function looks very much like run_all, they could easily be one function which accepts a `fail_fast` param; e.g.:
```
def run(nproc, fail_fast=False):
    cmds = [get_test_command(...), ...]
    error = False
    for cmd in cmds:
        status = run_cmd(cmd)
        if status != 0:
            if fail_fast:
                return status
            error = True
    return ERR_CODE if error else 0
```

Reply:
Done, but [this principle](https://www.informit.com/articles/article.aspx?p=1392524) kept me from doing this initially. In python at least you can name the parameter.

Reply:
Same thoughts as in the below case, but I will merely add, the function here does one thing: Run the test commands and return a status code; it just may not run *all* the commands, depending on the value of `fail_fast`. Anyways, this is a great example of principle-vs-principle being unhelpful: "DRY" vs. "boolean arguments are bad" gets you nowhere. The better questions to ask might be: "does this function make sense as a concept?"; "can I easily describe, e.g. in form of a comment, what this function does?"; "is the internal complexity of the function incredibly high, such that it needs an alternate composition?"; etc.

Reply:
The point of Uncle Bob here (as well in his book), is that whenever there is a boolean parameter, it suggests, that there should be 2 functions. 
But I'm not here to argue about principles, really. If the reviewers prefer it this way, I'm OK with this. Both approaches have pros and cons, but again, in Python at least you can name the boolean parameter when calling the function.